noir/noir-repo/aztec_macros/src/transforms/event_interface.rs

```diff
@@ -0,0 +1,393 @@
+            secondary_message: Some(format!("Failed to parse Noir macro code (trait impl of Serialize for {event_type}). This is either a bug in the compiler or the Noir macro code")),
+            span: Option::None
+        });
+    }
+
+    let mut sorted_ast = parsed_ast.into_sorted();
+    let serialize_impl = sorted_ast.trait_impls.remove(0);
+
+    Ok(serialize_impl)
+}
+
+fn generate_trait_impl_deserialize(
+    event_type: &str,
+    event_len: u32,
+    event_fields: &[(String, String)],
+) -> Result<NoirTraitImpl, AztecMacroError> {
+    let field_names: Vec<String> = event_fields
+        .iter()
+        .enumerate()
+        .map(|(index, field)| format!("{}: fields[{}]", field.0, index))
+        .collect::<Vec<String>>();
+    let field_input = field_names.join(",");
+
+    let trait_impl_source = format!(
+        "
+    impl dep::aztec::protocol_types::traits::Deserialize<{event_len}> for {event_type} {{
+        fn deserialize(fields: [Field; {event_len}]) -> {event_type} {{
+            {event_type} {{ {field_input} }}
+        }}
+    }}
+    "
+    )
+    .to_string();
+
+    let (parsed_ast, errors) = parse_program(&trait_impl_source);
+    if !errors.is_empty() {
+        dbg!(errors);
+        return Err(AztecMacroError::CouldNotImplementNoteInterface {
+            secondary_message: Some(format!("Failed to parse Noir macro code (trait impl of Deserialize for {event_type}). This is either a bug in the compiler or the Noir macro code")),
+            span: Option::None
+        });
+    }
+
+    let mut sorted_ast = parsed_ast.into_sorted();
+    let deserialize_impl = sorted_ast.trait_impls.remove(0);
+
+    Ok(deserialize_impl)
+}
+
+fn generate_fn_get_event_type_id(
+    event_type: &str,
+    field_length: u32,
+) -> Result<NoirFunction, AztecMacroError> {
+    let from_signature_input =
+        std::iter::repeat("Field").take(field_length as usize).collect::<Vec<_>>().join(",");
+    let function_source = format!(
+        "
+        fn get_event_type_id(self: {event_type}) -> dep::aztec::protocol_types::abis::event_selector::EventSelector {{
+           dep::aztec::protocol_types::abis::event_selector::EventSelector::from_signature(\"{event_type}({from_signature_input})\")
+    }}
+    ",
+    )
+    .to_string();
+
+    let (function_ast, errors) = parse_program(&function_source);
+    if !errors.is_empty() {
+        dbg!(errors);
+        return Err(AztecMacroError::CouldNotImplementNoteInterface {
+            secondary_message: Some(format!("Failed to parse Noir macro code (fn get_event_type_id, implemented for EventInterface of {event_type}). This is either a bug in the compiler or the Noir macro code")),
+            span: Option::None
+        });
+    }
+
+    let mut function_ast = function_ast.into_sorted();
+    let mut noir_fn = function_ast.functions.remove(0);
+    noir_fn.def.visibility = ItemVisibility::Public;
+    Ok(noir_fn)
+}
+
+fn generate_fn_selector(
```

Comment:
Do we need both a `selector` and a `get_event_type_id` for the event? Why do we have both? 👀